# 88. Merge Sorted Array

You are given two integer arrays nums1 and nums2, sorted in non-decreasing order, 
and two integers m and n, representing the number of elements in nums1 and nums2 
respectively.

Merge nums1 and nums2 into a single array sorted in non-decreasing order.

The final sorted array should not be returned by the function, but instead be 
stored inside the array nums1. To accommodate this, nums1 has a length of m + n, 
where the first m elements denote the elements that should be merged, and the l
ast n elements are set to 0 and should be ignored. nums2 has a length of n.

Example 1:

Input: nums1 = [1,2,3,0,0,0], m = 3, nums2 = [2,5,6], n = 3
Output: [1,2,2,3,5,6]
Explanation: The arrays we are merging are [1,2,3] and [2,5,6].
The result of the merge is [1,2,2,3,5,6] with the underlined elements coming 
from nums1.

#### Brute force solution

In [6]:
nums1 = [1,2,3,0,0,0]
m = 3
nums2 = [2,5,6]
n = 3

for i in range(m, m+n):
    nums1[i] = nums2[m-i]

nums1.sort()

print(nums1)


[1, 2, 2, 3, 5, 6]


#### Three Pointers (start from the end)
This is the most efficient solution

In [57]:



# Continue the loop while there are elements to compare in nums2 (p2 >= 0).
# The loop also checks (p >= 0) to ensure we don't go out of bounds in nums1.

def merge_sorted_arrays(nums1, nums2, m ,n):

    # Initialize read pointers for nums1 and nums2, starting from their end positions.
    p1 = m-1
    p2 = n-1

    # and writer point at the very end of num1
    p = m+n -1  
    
    while p >= 0 and p2 >=0 :

        if p1 > 0 and nums1[p1] > nums2[p2] :
            """
            
            """

            nums1[p] = nums1[p1]  # Place nums2's element in nums1 at location p.
            p1 = p1 - 1           # Move pointer p2 to the left in nums2.
        
        else:
            nums1[p] = nums2[p2] # Place nums1's element p1 in position p in nums1.
            p2 = p2 -1           # Move pointer p1 to the left in nums1.
        
        # Move write pointer `p` to the left to fill the next position.
        p = p-1 

nums1 = [1,2,3,0,0,0,0] 
nums2 = [2,5,6, 7]       
m, n = 3, 4

merge_sorted_arrays(nums1,nums2,m,n)

print(nums1)




[1, 2, 2, 3, 5, 6, 7]


# 238. Product of Array Except Self
% Topics: Array, Prefix Sum

Given an integer array nums, return an array answer such that answer[i] is equal
 to the product of all the elements of nums except nums[i].

The product of any prefix or suffix of nums is guaranteed to fit in a 32-bit 
integer.

You must write an algorithm that runs in O(n) time and without using the division 
operation.


Example 1:

Input: nums = [1,2,3,4]
Output: [24,12,8,6]
Example 2:

Input: nums = [-1,1,0,-3,3]
Output: [0,0,9,0,0]

 

In [11]:

# A) Brute Force
nums = [-1,1,0,-3,3]

result = []

for i in range(len(nums)):
    product = 1
    for j in range(len(nums)):
        if i != j:
            product = product * nums[j]
    
    result.append(product)

[0, 0, 9, 0, 0]


For every given index, i, we will make use of the product of all the numbers to 
the left of it and multiply it by the product of all the numbers to the right. 
This will give us the product of all the numbers except the one at the given 
index i.
<img src="../images/238_left.png" alt="left" width="450"/>
<img src="../images/238_right.png" alt="left" width="450"/>





1. We initialize lists left and right such that we can access and assign new 
values throughout the loop. THis is a usefull alternative than the usual append

2. Notice the second loop:  `for i in range(n - 2, -1, -1)`. This means that we 
go from n-2 to 0 taking steps of size 1, but in *reverse order*. 

In [39]:
# Left and Right Products
nums = [4, 5, 1, 8, 2, 10, 6]
n = len(nums)

left = [1] * n
right = [1] * n

for i in range(1, n): 
    left[i] = left[i - 1] * nums[i - 1]
    #print(f"i:{i}, left:{left}, nums:{nums} ")

for i in range(n - 2, -1, -1):
    right[i] = right[i + 1] * nums[i + 1]
    #print(f"i:{i}, right:{right}, nums:{nums} ")
answer = []
for i in range(n):
    product = left[i] * right[i]
    answer.append(product)
     
print(f"answer: {answer}")


i:1, left:[1, 4, 1, 1, 1, 1, 1], nums:[4, 5, 1, 8, 2, 10, 6] 
i:2, left:[1, 4, 20, 1, 1, 1, 1], nums:[4, 5, 1, 8, 2, 10, 6] 
i:3, left:[1, 4, 20, 20, 1, 1, 1], nums:[4, 5, 1, 8, 2, 10, 6] 
i:4, left:[1, 4, 20, 20, 160, 1, 1], nums:[4, 5, 1, 8, 2, 10, 6] 
i:5, left:[1, 4, 20, 20, 160, 320, 1], nums:[4, 5, 1, 8, 2, 10, 6] 
i:6, left:[1, 4, 20, 20, 160, 320, 3200], nums:[4, 5, 1, 8, 2, 10, 6] 
i:5, right:[1, 1, 1, 1, 1, 6, 1], nums:[4, 5, 1, 8, 2, 10, 6] 
i:4, right:[1, 1, 1, 1, 60, 6, 1], nums:[4, 5, 1, 8, 2, 10, 6] 
i:3, right:[1, 1, 1, 120, 60, 6, 1], nums:[4, 5, 1, 8, 2, 10, 6] 
i:2, right:[1, 1, 960, 120, 60, 6, 1], nums:[4, 5, 1, 8, 2, 10, 6] 
i:1, right:[1, 960, 960, 120, 60, 6, 1], nums:[4, 5, 1, 8, 2, 10, 6] 
i:0, right:[4800, 960, 960, 120, 60, 6, 1], nums:[4, 5, 1, 8, 2, 10, 6] 
answer: [4800, 3840, 19200, 2400, 9600, 1920, 3200]


# 7. Reverse Integer

Theme: math.
Given a signed 32-bit integer x, return x with its digits reversed. If reversing 
x causes the value to go outside the signed 32-bit integer range [-231, 231 - 1],
then return 0.

* Example 1: Input: x = 123. | Output: 321

* Example 2:  Input: x = -123 | Output: -321

* Example 3: Input: x = 120 | Output: 21

In [102]:
class Solution:
    def reverse(self, x:int) -> int:
        if x > 0:
            x_str = str(x)
            x_str = x_str[::-1]
            x_rvd = int(x_str)
        else:
            x_str = str(x*(-1))
            x_str = x_str[::-1]
            x_rvd = int(x_str)*(-1)

        return x_rvd if x_rvd > (-2**31) and x_rvd < 2 ** 31 else 0

x = -432

Solution().reverse(x)

-234

### Slicing in Python

Notice that we use the slice [::-1] in the solution above. Here is a quick 
summary of slicing in Python:

Slicing allows you to access a subset of elements from sequences such as strings, 
lists, tuples, and other objects that support the sequence protocol.

#### Basic Syntax
`sequence[start:stop:step]``
	• start: The index where the slice starts (inclusive). If omitted, the 
	slice starts from the beginning.
	• stop: The index where the slice ends (exclusive). If omitted, the slice 
	goes up to the end of the sequence.
	• step: The step size or stride. If omitted, it defaults to 1. A negative
	 step reverses the direction.

`my_list = [0, 1, 2, 3, 4, 5]`

1. Basic Slicing	
	* `slice1 = my_list[1:4]`  
	Output: [1, 2, 3 ]

2. Omitting start or stop
	* `slice2 = my_list[:3]`  
	 Output: [0, 1, 2]
	* `slice3 = my_list[3:]`  
	 Output: [3, 4, 5]

3. Using steps to control the stride
	* Every second elevent
	`slice5 = my_list[::2]`
	Output: [0, 2, 4]

4. Steps = -1 to reverse the sequence
	* `reversed_list = my_list[::-1]`  
	outputs [5, 4, 3, 2, 1, 0]

####  9. Palindrome Number
Given an integer x, return true if x is a 
palindrome, and false otherwise.

Input: x = 121 | Output: true
Explanation: 121 reads as 121 from left to right and from right to left.


Input: x = -121 | Output: false
Explanation: From left to right, it reads -121. From right to left, it becomes 121-. Therefore it is not a palindrome.

In [152]:
class Solution:
    def isPalindrome(self, x:int) -> bool:
        x_str = str(x)
        x_rvd = x_str[::-1]
        x_rvd

        if x_rvd == x_str:
            return True
        else:
            return False

Not a palindrome!


In [ ]:
def isPossible(self, A):
    """
    Determines if it's possible to split the sorted integer array 'A' into one or more subsequences
    such that each subsequence is a consecutive increasing sequence of length at least 3.

    Parameters:
    - A: List[int] - A sorted list of integers.

    Returns:
    - bool: True if possible to split as per the conditions, False otherwise.
    """
    import collections  # Import the 'collections' module to use 'Counter'.

    # 'left' counts the number of times each number appears in 'A'.
    # It represents the numbers that are still available to be placed in subsequences.
    left = collections.Counter(A)

    # 'end' counts the number of subsequences ending with a specific number.
    # It helps track subsequences that can be extended.
    end = collections.Counter()

    # Iterate over each number 'i' in the array 'A'.
    for i in A:
        # If 'i' has already been used up in subsequences, skip it.
        if not left[i]:
            continue

        # Use one occurrence of 'i'.
        left[i] -= 1

        # Check if there's a subsequence ending with 'i - 1' that can be extended.
        if end[i - 1] > 0:
            # Decrement the count of subsequences ending with 'i - 1'.
            end[i - 1] -= 1
            # Increment the count of subsequences now ending with 'i'.
            end[i] += 1
        # If no such subsequence exists, try to create a new subsequence [i, i+1, i+2].
        elif left[i + 1] > 0 and left[i + 2] > 0:
            # Use one occurrence of 'i + 1'.
            left[i + 1] -= 1
            # Use one occurrence of 'i + 2'.
            left[i + 2] -= 1
            # Increment the count of subsequences ending with 'i + 2'.
            end[i + 2] += 1
        else:
            # Cannot place 'i' in any subsequence as per the rules.
            return False  # Early exit since the conditions cannot be met.

    # All numbers have been placed successfully in valid subsequences.
    return True

### 659. Split Array into Consecutive Subsequences
https://leetcode.com/problems/split-array-into-consecutive-subsequences/description/?envType=company&envId=google&favoriteSlug=google-all&difficulty=EASY%2CMEDIUM&role=data-scientist-data-engineer

Medium / hard

You are given an integer array nums that is **sorted in non-decreasing order.**

Determine if it is possible to split nums into **one or more subsequences** such
that both of the following conditions are true:

1. Each subsequence is a consecutive increasing sequence (i.e. each integer is 
exactly one more than the previous integer).
2. All subsequences have a length of 3 or **more**.

Return true if you can split nums according to the above conditions, or false
otherwise.

A **subsequence** of an array is a new array that is formed from the original
array by deleting some (can be none) of the elements without disturbing the 
relative positions of the remaining elements. (i.e., [1,3,5] is a subsequence of
 [1,2,3,4,5] while [1,3,2] is not).

<img src="../images/659_1.png" alt="Left" width="450"/>
<img src="../images/659_2.png" alt="right" width="450"/>
<img src="../images/659_3.png" alt="right" width="450"/>
<img src="../images/659_4.png" alt="right" width="450"/>
<img src="../images/659_5.png" alt="right" width="450"/>

In [12]:
def isPossible(self, nums):
    """
    Determines if it's possible to split the sorted integer array 'nums' into one or more subsequences
    such that each subsequence is a consecutive increasing sequence of length at least 3.

    Parameters:
    - nums: List[int] - nums sorted list of integers.

    Returns:
    - bool: True if possible to split as per the conditions, False otherwise.
    """
    import collections  # Import the 'collections' module to use 'Counter'.

    # 'frequency_map' counts the number of times each number appears in 'nums'.
    # It represents the numbers that are still available to be placed in subsequences.
    frequency_map = collections.Counter(nums)

    # 'end' counts the number of subsequences ending with a specific number.
    # It helps track subsequences that can be extended.
    subsequence_map = collections.Counter()

    # Iterate over each number 'i' in the array 'nums'.
    for i in nums:
        # If 'i' has already been used up in subsequences, skip it.
        if not frequency_map[i]:
            continue
        
        frequency_map[i] -= 1    # Use one occurrence of 'i'.

        # Check if there's a subsequence ending with 'i - 1' that can be extended.
        if subsequence_map[i - 1] > 0:
            subsequence_map[i - 1] -= 1 # Decrement the count of subsequences ending with 'i - 1'.
            subsequence_map[i] += 1      # Increment the count of subsequences now ending with 'i'.

        # If no such subsequence exists, try to create a new subsequence [i, i+1, i+2].
        elif frequency_map[i + 1] > 0 and frequency_map[i + 2] > 0:
            frequency_map[i + 1] -= 1 # Use one occurrence of 'i + 1' 
            frequency_map[i + 2] -= 1 # Use one occurrence of 'i + 2'
            subsequence_map[i + 2] += 1 # Increment the count of subsequences ending with 'i + 2'.

        else:
            # Cannot place 'i' in any subsequence as per the rules.
            return False  # Early exit since the conditions cannot be met.

    # All numbers have been placed successfully in valid subsequences.
    return True